substrate/frame/transaction-payment/src/lib.rs

```diff
@@ -683,6 +695,59 @@ impl<T: Config> Pallet<T> {
 	pub fn deposit_fee_paid_event(who: T::AccountId, actual_fee: BalanceOf<T>, tip: BalanceOf<T>) {
 		Self::deposit_event(Event::TransactionFeePaid { who, actual_fee, tip });
 	}
+
+	/// Withdraw `amount` from the currents transaction's fees.
+	///
+	/// If enough balance is available a credit of size `amount` is returned.
+	///
+	/// # Warning
+	///
+	/// This is only useful if a pallet knows that the pre-dispatch weight was vastly
```

Comment:
I think that the condition here is more subtle: the pallet is allowed to draw too much balance but would need to fail the transaction if that is the case (this more relaxed condition is required for pallet-revive).

It is important that in this case any effects that the balance was drawn for during execution (e.g., storage deposits) are reverted – in order to protect against DoS.

Reply:
The documentation still not expresses that this is a possible footgun and that you need to pay extreme attention to what you are doing. Because, just rolling back changes, may still results in a lot of computation not being paid for. The code using this function needs to ensure that it tracks how much weight itself registered and only withdraws that much fee. Maybe it would be even better to withdraw using `Weight` directly? And then internally convert to a balance?

Reply:
Added an additional warning that this should not be used for weight fees.